### PR TITLE
Text-to-Text Finetuning 

### PR DESCRIPTION
**Pull request by EricFillion**
Allows users to fine-tune models for text-to-text tasks, such as translation, grammar correction, summarization and more. 

Hugging Face created special components for finetuning text-to-text models. For example, instead of using the class called "TrainingArguments()" they recommend using "Seq2SeqTrainingArguments()." Similarly, instead of using the "Trainer()" class, they recommend using "Seq2SeqTrainer()." So, the methods within Trainer() class were not used as this functionality requires custom components. 

**Potential Changes** 
1. Change the header values for training/evaluating. Currently they are "text_1" and "text_2." 
2. "TTTrainArgs()" and "TTTestArgs()" sound awkward. But, in my opinion,  consistency is more important than a couple of awkward-sounding names. 

**Next Steps** 
_These can be accomplished in a later release_ 
1. Saving/loading processed data. 
2. HappyTextToText.test() 
3. Other evaluation metrics like rouge. But it may be better to explain to users how to compute various evaluation metrics externally to minimize dependencies and avoid adding too many complexities. 

**Other Notes** 
1. I did not include the ability to use a dictionary as the input for the  _args_ parameter for HappyTextToText()'s train and eval methods. We added the ability to use dataclass objects for train/eval args parameters in version 2.2.0 and have been recommending them over dictionaries ever since for all training functionality. So, going forward, we will stop adding the ability to use dictionaries as inputs for new methods.

**Resources** 
1. Documentation for the features can be found [here]( https://github.com/EricFillion/happy-transformer/blob/ef/t-t-finetuning/docs/pages/7-text-to-text/4-finetuning.md)
2. A full example can be found [here](https://github.com/EricFillion/happy-transformer/blob/ef/t-t-finetuning/examples/text-to-text/training_summarization.py)
3. [Here's](https://github.com/huggingface/notebooks/blob/master/examples/summarization.ipynb) one of the main sources this PR is based on
4. Run these new features in [Colab](https://colab.research.google.com/drive/1-SJFZgJDTD00qbxwdkTR8vYcPyJxHlU4?usp=sharing)

### CODE REVIEW

**Comment by EricFillion** [comment]
I have it working for three unique tasks: [summarization](https://github.com/EricFillion/happy-transformer/blob/ef/t-t-finetuning/examples/text-to-text/training_summarization.py), [grammar correction](https://github.com/EricFillion/happy-transformer/blob/ef/t-t-finetuning/examples/text-to-text/training_grammar.py) and [translation](https://github.com/EricFillion/happy-transformer/blob/ef/t-t-finetuning/examples/text-to-text/training_translation.py).

File: happytransformer/gen/trainer.py [hunk 1]
```diff
@@ -72,6 +72,7 @@ def train(self, input_filepath, dataclass_args: GENTrainArgs):
             self._generate_json(dataclass_args.save_preprocessed_data_path, tokenized_dataset["train"], "train")
 
         self.logger.info("Training...")
+        print("tokenized_dataset['train']", tokenized_dataset['train'])
```
Comment (ted537) [inline happytransformer/gen/trainer.py:75]: Printing straight to `stdout` as a side effect in a library feels wrong

File: happytransformer/happy_text_to_text.py [hunk 2]
```diff
@@ -6,18 +6,26 @@
 from transformers import Text2TextGenerationPipeline, AutoModelForSeq2SeqLM
 
 from happytransformer.happy_transformer import HappyTransformer
-from happytransformer.tt.trainer import TTTrainer
+from happytransformer.tt.trainer import TTrainer
```
Comment (ted537) [inline happytransformer/happy_text_to_text.py:9]: Wouldn't `TTTrainer` be more in line with the other types?

File: examples/text-to-text/training_grammar.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        writter = csv.writer(csvfile)
+        writter.writerow(["text_1", "text_2"])
+        for case in dataset:
+            input_text = "grammar: " + case["sentence"]
+            for correction in case["corrections"]:
+                # a few of the case are have None values. We'll skip them
+                if not input_text or not correction:
+                    pass
+                else:
+                    writter.writerow([input_text, correction])
```
[the remark this answers is not in the document]
Comment (EricFillion) [reply examples/text-to-text/training_grammar.py]: Updated: 
b753ac1

File: happytransformer/gen/trainer.py (hunk 1 above)
Comment (EricFillion) [reply happytransformer/gen/trainer.py]: Forgot to delete that: 
a211655

File: happytransformer/happy_text_to_text.py (hunk 2 above)
Comment (EricFillion) [reply happytransformer/happy_text_to_text.py]: I agree 
522ed3e

**Comment by ted537** [comment]
I agree that "text_1" and "text_2"  in the CSV format should be reconsidered. It's not immediately clear what they represent.

**Comment by EricFillion** [comment]
> I agree that "text_1" and "text_2" in the CSV format should be reconsidered. It's not immediately clear what they represent.

Perhaps "input_text' and "output_text." Or, "input_text' and "target_text."

**Comment by EricFillion** [comment]
> I agree that "text_1" and "text_2" in the CSV format should be reconsidered. It's not immediately clear what they represent.

Hugging Face uses the terms input/target so maybe we should go with "input_text" or "target_text" or even just "input" and "target"

**Comment by ted537** [comment]
Yeah I agree with "input" and "target". The contents of the rest of the file will make it very obvious that these columns are describing text.

**Review by ted537 — APPROVED** [review]

**Pull request merged**